validator/src/journal/block_manager.rs

```diff
@@ -398,13 +402,158 @@ impl BlockManagerState {
 #[derive(Default, Clone)]
 pub struct BlockManager {
     state: Arc<BlockManagerState>,
+    persist_lock: Arc<RwLock<()>>,
 }
 
 impl BlockManager {
     pub fn new() -> Self {
         BlockManager::default()
     }
 
+    fn transaction_index_contains(
+        &self,
+        index: &Box<IndexedBlockStore>,
+        block_id: &str,
+        id: &str,
+    ) -> Result<bool, BlockManagerError> {
+        let index = index.as_ref();
+        if TransactionIndex::contains(index, id)? {
+            let block = TransactionIndex::get_block_by_id(index, id)?
+                .expect("BlockStore updated during transaction index check");
+            let head = index
+                .get(&[block_id])?
+                .next()
+                .expect("BlockStore updated during transaction index check");
+            if block.block_num <= head.block_num {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        } else {
+            Ok(false)
+        }
+    }
+
+    fn batch_index_contains(
+        &self,
+        index: &Box<IndexedBlockStore>,
+        block_id: &str,
+        id: &str,
+    ) -> Result<bool, BlockManagerError> {
+        let index = index.as_ref();
+        if BatchIndex::contains(index, id)? {
+            let block = BatchIndex::get_block_by_id(index, id)?
+                .expect("BlockStore updated during batch index check");
+            let head = index
+                .get(&[block_id])?
+                .next()
+                .expect("BlockStore updated during batch index check");
+            if block.block_num <= head.block_num {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        } else {
+            Ok(false)
+        }
+    }
+
+    pub fn contains_transaction(
+        &self,
+        block_id: &str,
+        id: &str,
+    ) -> Result<bool, BlockManagerError> {
+        let _lock = self
+            .persist_lock
+            .read()
+            .expect("The persist RwLock is poisoned");
+        if let Some((_, store)) = self
+            .state
+            .blockstore_by_name
+            .read()
+            .expect("Blockstore RW Lock is poisoned")
+            .iter()
+            .find(|(_, store)| store.get(&[block_id]).map(|res| res.count()).unwrap_or(0) > 0)
+        {
+            self.transaction_index_contains(store, block_id, id)
```

Comment:
I don't think there is any guarantee here that block_id won't be dropped from the store in between when the above get is called and this line executes. I think you need to hold the lock around `block_by_block_id` for the duration of this function.